fri/src/prover.rs

```diff
@@ -95,20 +96,57 @@ where
     let mut data = vec![];
 
     while folded.len() > config.blowup() * config.final_poly_len() {
-        let leaves = RowMajorMatrix::new(folded, 2);
-        let (commit, prover_data) = config.mmcs.commit_matrix(leaves);
+        let arity = config.arity();
+        let next_folded_len = folded.len() / arity;
+
+        // First, we collect the polynomial evaluations that will be committed this round.
+        // Those are `folded` and polynomials in `inputs` not consumed yet with number of
+        // evaluations more than `next_folded_len`
+        let mut polys_before_next_round = vec![];
+
+        let mut cur_folded_len = folded.len();
+        while cur_folded_len > next_folded_len {
+            if let Some(poly_eval) = inputs_iter.next_if(|v| v.len() == cur_folded_len) {
+                let width = poly_eval.len() / next_folded_len;
+                let poly_eval_matrix = RowMajorMatrix::new(poly_eval, width);
+                polys_before_next_round.push(poly_eval_matrix);
+            }
+
+            cur_folded_len /= 2;
+        }
+
+        let folded_matrix = RowMajorMatrix::new(folded.clone(), arity);
+        let matrices_to_commit: Vec<DenseMatrix<Challenge>> = iter::once(folded_matrix)
+            .chain(polys_before_next_round)
+            .collect();
+
+        let (commit, prover_data) = config.mmcs.commit(matrices_to_commit);
         challenger.observe(commit.clone());
 
+        // Next, we fold `folded` and `polys_before_next_round` to prepare for the next round
         let beta: Challenge = challenger.sample_ext_element();
-        // We passed ownership of `current` to the MMCS, so get a reference to it
-        let leaves = config.mmcs.get_matrices(&prover_data).pop().unwrap();
-        folded = g.fold_matrix(beta, leaves.as_view());
+
+        // Get a reference to the committed matrices
+        let leaves = config.mmcs.get_matrices(&prover_data);
+        let mut leaves_iter = leaves.into_iter().peekable();
+        // Skip `folded`
+        leaves_iter.next();
+
+        while folded.len() > next_folded_len {
+            let matrix_to_fold = RowMajorMatrix::new(folded, 2);
+            folded = g.fold_matrix(beta, matrix_to_fold);
```

Comment:
square `beta` after each fold

Reply:
I don't think there is a need for that. `fold_matrix` figures out the correct powers based on the height of `matrix_to_fold`

Reply:
so the issue we had discussed before is resolved? around folding twice vs folding arity=4 once

Reply:
I think so, although it would be nice if you can look at `fold_matrix` to make sure

Reply:
Nevermind, you're right. Fixed